product_tax_multicompany_default/models/product.py

```diff
@@ -0,0 +1,258 @@
+            ).ids
+            candidates = set()
+            all_companies = all_companies.filtered(lambda c: c.id != current_company.id)
+            for company in all_companies:
+                candidates.update(
+                    self._taxes_by_company(field_name_data, company, current_tax_ids)
+                )
+            if len(candidates) > 0:
+                propagate_taxes = True
+        else:
+            current_names = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).mapped("name")
+            current_tax_ids = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).ids
+            for company_id, value in taxes_bd.items():
+                if current_names == list(value.values()):
+                    continue
+                # We are looking to see if there are any taxes that can be applied to
+                # other companies from the taxes that the current product has.
+                candidates = self._taxes_by_company(
+                    field_name_data,
+                    self.env["res.company"].browse(company_id),
+                    current_tax_ids,
+                )
+                if len(candidates) == len(current_tax):
+                    propagate_taxes = True
+                    break
+
+        return propagate_taxes
+
+    def _get_product_taxes(self, field):
+        """Returns taxes on purchase or sales taxes
+
+        We need the product taxes for other companies, and we cannot take
+        them from the object itself, so we have to consult the database
+        where the product taxes passed by parameter are located for all
+        companies.
+
+        :param field: can be 'taxes' or 'purchase'
+        """
+        self.ensure_one()
+
+        if not self.id:
+            return []
+
+        table_by_field = {
+            "taxes": "product_taxes_rel",
+            "purchase": "product_supplier_taxes_rel",
+        }
+        table_name = table_by_field.get(field)
+
+        if not table_name:
+            raise ValueError("field should be 'taxes' or 'purchase'")
+
+        sql = f"""
+            SELECT DISTINCT a_tax.company_id, a_tax.name
+            FROM {table_name} ptr
+            LEFT JOIN account_tax a_tax ON tax_id = a_tax.id
+            WHERE ptr.prod_id = %s
+            """
+        self.env.cr.execute(sql, [self._origin.id])
+        return self.env.cr.fetchall()
+
+    def _taxes_by_company(self, field, company, match_tax_ids=None):
+        taxes_ids = []
+        if match_tax_ids is None:
+            taxes_ids = company[field].ids
+        # If None: return default taxes
+        if not match_tax_ids:
+            return taxes_ids
+        type_tax_use = "sale" if field == "account_sale_tax_id" else "purchase"
+        account_tax = self.env["account.tax"].sudo().browse(match_tax_ids)
+        taxes_ids = account_tax.filtered_domain([("type_tax_use", "=", type_tax_use)])
```

Comment:
Rename to a word that doesn't contain _ids. This name looks like the variable conatains a list of ids instead of records.

```suggestion
        all_taxes = self.env["account.tax"].sudo().browse(match_tax_ids)
        type_use_taxes = account_tax.filtered_domain([("type_tax_use", "=", type_tax_use)])
```